drivers/sms_ser.c

```diff
@@ -0,0 +1,701 @@
+
+    for (int i = 1; i < BUFFER_SIZE - 2; i++) {
+        if (i <= 12) {
+            snprintfcat(results->model, 24, "%c", rawvalues[i]);
+        } else {
+            snprintfcat(results->version, 6, "%c", rawvalues[i]);
+        }
+    }
+}
+
+void sms_parse_results(uint8_t *rawvalues, SmsData *results) {
+    char buf[BUFFER_SIZE];
+    uint8_t byte, mask;
+    long v;
+    double h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[1], (unsigned char)rawvalues[2]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->lastinputVac = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[3], (unsigned char)rawvalues[4]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->inputVac = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[5], (unsigned char)rawvalues[6]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->outputVac = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[7], (unsigned char)rawvalues[8]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->outputpower = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[9], (unsigned char)rawvalues[10]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->outputHz = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[11], (unsigned char)rawvalues[12]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->batterylevel = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[13], (unsigned char)rawvalues[14]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->temperatureC = h;
+
+    byte = rawvalues[15];
+    mask = 1;
+
+    results->beepon = ((byte & (mask << 0)) != 0) ? true : false;
+    results->shutdown = ((byte & (mask << 1)) != 0) ? true : false;
+    results->test = ((byte & (mask << 2)) != 0) ? true : false;
+    results->upsok = ((byte & (mask << 3)) != 0) ? true : false;
+    results->boost = ((byte & (mask << 4)) != 0) ? true : false;
+    results->bypass = ((byte & (mask << 5)) != 0) ? true : false;
+    results->lowbattery = ((byte & (mask << 6)) != 0) ? true : false;
+    results->onbattery = ((byte & (mask << 7)) != 0) ? true : false;
+}
+
+static int get_ups_nominal() {
+    uint8_t length;
+    ssize_t ret;
+
+    upsdebugx(LOG_DEBUG, "get_ups_nominal");
+
+    length = sms_prepare_get_status(&bufOut[0]);
+
+    if (ser_send_buf(upsfd, bufOut, length) == 0) {
+        upsdebugx(LOG_ERR, "Communication error while writing to port");
+        return -1;
+    }
+    memset(bufIn, 0, BUFFER_SIZE);
+    ret = ser_get_buf_len(upsfd, &bufIn[0], BUFFER_SIZE, 3, 1000);
```

Comment:
As a bit of coding style for future, it is more future-proof to pass size constraints as `sizeof(arrayname)` than as the specific number (or macro) that happened to initialize that size in some iteration of code.